core/sql/optimizer/PartFunc.cpp

```diff
@@ -5009,6 +5020,118 @@ void RangePartitioningFunction::print(FILE* ofd, const char* indent,
   partitionBoundaries_->print(ofd, indent, title);
 
 } // RangePartitioningFunction::print()
+
+
+NABoolean 
+compareEncodedKey(const char* low, const char* key, const char* high, Int32 keyLen, NABoolean checkLast)
+{
+    Int32 cmpLow = memcmp(low, key, keyLen);
+    Int32 cmpHigh = memcmp(key, high, keyLen);
+
+    if ( cmpLow <= 0 && cmpHigh < 0 )
+       return TRUE;
+
+    return (checkLast && cmpLow <= 0 && cmpHigh <= 0);
+}
+
+NABoolean 
+compareAsciiKey(const char* low, const char* key, const char* high, Int32, NABoolean checkLast)
+{
+    Int32 cmpLow = strverscmp(low, key);
+    Int32 cmpHigh = strverscmp(key, high);
+
+    if ( cmpLow <= 0 && cmpHigh < 0 )
+       return TRUE;
+
+    return (checkLast && cmpLow <= 0 && cmpHigh <= 0);
+}
+
+
+// find a boundary pair [low, high) with smallest low value in which keys fall, and return the
+// index of the boundary low. Return -1 otherwise, or the key lengths are different.
+Int32 RangePartitionBoundaries::findBeginBoundary(char* encodedKey, Int32 keyLen, 
+                                                  compFuncPtrT compFunc) const
+{
+   // boundaries are stored in entries in the range [0, partitionCount_] 
+   for (Lng32 i=partitionCount_-1; i>= 0; i--) {
+
+       const char* low = getBinaryBoundaryValue(i);
+       const char* high = getBinaryBoundaryValue(i+1);
+
+       // test if encodedKey is in [low, high)
+       if ( (*compFunc)(low, encodedKey, high, keyLen, i==partitionCount_-1) )
+          return i;
+   }
+
+   return -1;
+}
+
+// find a boundary pair [low, high) with the largest low value in which keys fall, and return the
+// index of the boundary low. Return -1 otherwise, or the key lengths are different.
+Int32 RangePartitionBoundaries::findEndBoundary(char* encodedKey, Int32 keyLen, 
+                                                compFuncPtrT compFunc) const
+{
+   // boundaries are stored in entries in the range [0, partitionCount_] 
+   for (Lng32 i=0; i<partitionCount_-1; i++ ) {
+
+       const char* low = getBinaryBoundaryValue(i);
+       const char* high = getBinaryBoundaryValue(i+1);
+
+       // test if encodedKey is in [low, high)
+       if ( (*compFunc)(low, encodedKey, high, keyLen, i==partitionCount_-1) )
+          return i;
+   }
+
+   return -1;
+}
+
+Int32 
+RangePartitioningFunction::computeNumOfActivePartitions(SearchKey* skey, const TableDesc* tDesc) const
+{
+   const RangePartitionBoundaries* boundaries = getRangePartitionBoundaries();
+
+   Int32 origPartitions = getCountOfPartitions();
+   Int32 partitions = origPartitions;
+   Int32 bIndex = 0;
+
+   const NATable* naTable = tDesc->getNATable();
+   NABoolean isNativeHbase = (naTable->isHbaseCellTable() || naTable->isHbaseRowTable());
+   compFuncPtrT compFuncPtr = ( isNativeHbase ) ? compareAsciiKey: compareEncodedKey;
```

Comment:
Good point. I'll add a check for hive.

On Mon, Oct 19, 2015 at 1:30 PM, DaveBirdsall <notifications@github.com>
wrote:

> In core/sql/optimizer/PartFunc.cpp
> <https://github.com/apache/incubator-trafodion/pull/132#discussion_r42408115>
> :
>
> > +
> > +   return -1;
> > +}
> > +
> > +Int32
> > +RangePartitioningFunction::computeNumOfActivePartitions(SearchKey* skey, const TableDesc* tDesc) const
> > +{
> > +   const RangePartitionBoundaries* boundaries = getRangePartitionBoundaries();
> > +
> > +   Int32 origPartitions = getCountOfPartitions();
> > +   Int32 partitions = origPartitions;
> > +   Int32 bIndex = 0;
> > +
> > +   const NATable* naTable = tDesc->getNATable();
> > +   NABoolean isNativeHbase = (naTable->isHbaseCellTable() || naTable->isHbaseRowTable());
> > +   compFuncPtrT compFuncPtr = ( isNativeHbase ) ? compareAsciiKey: compareEncodedKey;
>
> Do we care about Hive tables in this code path?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/132/files#r42408115>.
>



-- 
Regards, --Qifan